src/maxminddb/lib.rs

```diff
@@ -110,6 +115,89 @@ impl<'de> Reader<Vec<u8>> {
     }
 }
 
+#[derive(Debug)]
+struct WithinNode {
+    node: usize,
+    ip_bytes: Vec<u8>,
+    prefix_len: usize,
+}
+
+#[derive(Debug)]
+pub struct Within<'de, T: Deserialize<'de>, S: AsRef<[u8]>> {
+    reader: &'de Reader<S>,
+    node_count: usize,
+    stack: Vec<WithinNode>,
+    phantom: PhantomData<&'de T>,
+}
+
+#[derive(Debug)]
+pub struct WithinItem<T> {
+    pub ip_net: IpNetwork,
+    pub info: T,
+}
+
+impl<'de, T: Deserialize<'de>, S: AsRef<[u8]>> Iterator for Within<'de, T, S> {
+    type Item = Result<WithinItem<T>, MaxMindDBError>;
```

Comment:
Fairly new to rust (in case it's not obvious) and I'm not finding a clean way to go about returning errors that happen during the walking of nodes. My searching for idiomatic patterns didn't turn anything up, but it's a tough thing to search for.

This results in some pretty ugly code below to map errors in the helper functions into `Some(Err(e))`. That seems bad enough, but currently have have things in `match` `Ok`/`Err` conditionals that make it tough to read. Things like `.map_err` won't let me wrap with `Some` so I don't see a cleaner way. Seems like either a shortcoming of iterators or I'm just going about this all wrong :grin:

Input/thoughts/suggestions welcome.

Reply:
Looks like there may be a crate, https://docs.rs/fallible-iterator/0.2.0/fallible_iterator/ that calls out the situation. 

Seems like using it might be the cleaner way to go, but will defer to more experienced opinions.

Reply:
I have a `FallibleIterator` based solution and it cleans up pretty much all of the ugliness in error handling here. It's not perfect since you can't use `for x in y`, and have to go with `while let Some(x) = y.next()`, but that's a limitation of rust and seems like a reasonable tradeoff for cleaner error handling. 

Changes for that can be seen with https://github.com/ross/maxminddb-rust/compare/within-support...ross:within-support-fallibleiterator?expand=1. Thoughts welcome.

Reply:
Although it does clean up the code somewhat, I think I'd prefer not having the dependency, especially given that the fallible iterator crate hasn't seen an update in two years and doesn't seem that widely used.

Reply:
I think it hasn't seen any updates b/c it does what it needs to do. Similar to how I wrote requests-futures 8 years ago and haven't really touched it meaningfully in most of that time. 

As for not wanting the dep, I don't have a response to that. The code to implement the non-fallible iterator is ugly as hell, but using it isn't THAT bad. Kind of seems like a shortcoming of Rust itself that iterators can't fail.  

I'll stick this back on my TODO list to pick back up and see if I can get it over the line.